requests-for-comments/2025-04-15-scripts-to-rule-them-all.md

```diff
@@ -0,0 +1,72 @@
+# Request for comments: Consistent scripts to rule them all
+
+## Problem statement
+
+*Who are we building for, what are their needs (include example cases), why is this important?*
+
+PostHog has 235 (at the time of writing) public repositories. Each of these repositories has a unique way to get the project up and running locally.
+Each repository has a unique way to run tests, build the project, and so on. For those of us who work on the many client SDK libraries, this adds a
+lot of friction as we switch from project to project and have to learn a whole new set of incantations. Not only that, it makes it more challenging
+for external contributors to get involved as they have to learn a whole new process for each repository they work on.
+
+And as we continue to grow, this challenge will only grow.
+
+## Solution
+
+Adopt and adapt GitHub's approach of [scripts to rule them all](https://github.blog/engineering/engineering-principles/scripts-to-rule-them-all/) to
+fit our needs.
+
+> Being able to get from git clone to an up-and-running project in a development environment is imperative for fast, reliable contributions.
+
+I'm not suggesting we do _exactly_ what they do, but a similar approach would be a big benefit. Not every repository will need every script. Some
+repositories will need scripts custom to the environment. That's all fine. The goal is to have a baseline set of scripts that we can use to get a
+project up and running in a development environment and a known location to look for those scripts.
+
+## Context
+
+For context, here's the list of scripts referenced in the GitHub blog post ([samples available in a repository](https://github.com/github/scripts-to-rule-them-all)):
+
+* `script/bootstrap` - Install dependencies (Ex. npm packages, brew packages, etc. Usually run once after cloning the repository)
+* `script/setup` - Setup the project. Seems like this could just be merged in bootstrap
+* `script/update` - Update project after a pull such as migrations, etc.
+* `script/server` - For web apps, starts the application. For client SDKs, might start an example site.
+* `script/test` - Run tests (Ex. `npm test`, `bundle exec rspec`, etc.). Also includes linting, formatting, etc.
+* `script/cibuild` - used by the CI system to run tests, build, etc.
+* `script/console` - is used to open a console for your application.
+
+## Design
```

Comment:
Love the idea!

I'd love to also standardise the release process for new versions of a library, I don't think you need to do that in this PR though

Reply:
> I'd love to also standardise the release process for new versions of a library, I don't think you need to do that in this PR though

Same! It's on my list of things to do. With RELEASING, I think I'll document the many ways we do it today and propose a couple ideas for how we can consolidate it into a common approach. Obviously, not all things can be released in exactly the same way, but we can probably come up with a framework for releasing things that's fairly consistent.

Reply:
You can inspire from (or even use it) https://github.com/getsentry/craft
We built craft at sentry exactly for this reason

Reply:
I'd also recommend having the release and contributing markdown in the repos
eg https://github.com/PostHog/posthog-android/blob/main/RELEASING.md
https://github.com/PostHog/posthog-android/blob/main/CONTRIBUTING.md
so that open source contributors (and new employees) know how to build, run, and release things.